src/Analyzers/CSharp/Analyzers/NewLines/ConditionalExpressionPlacement/ConditionalExpressionPlacementDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.NewLines.ConditionalExpressionPlacement
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class ConditionalExpressionPlacementDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public ConditionalExpressionPlacementDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.ConditionalExpressionPlacementDiagnosticId,
+                   EnforceOnBuildValues.ConsecutiveBracePlacement,
+                   CSharpCodeStyleOptions.AllowBlankLineAfterTokenInConditionalExpression,
+                   new LocalizableResourceString(
+                       nameof(CSharpAnalyzersResources.Blank_line_not_allowed_after_conditional_expression_token), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeTree);
+
+        private void AnalyzeTree(SyntaxTreeAnalysisContext context)
+        {
+            var option = context.GetCSharpAnalyzerOptions().AllowBlankLineAfterTokenInConditionalExpression;
+            if (option.Value)
+                return;
+
+            Recurse(context, option.Notification.Severity, context.Tree.GetRoot(context.CancellationToken));
```

Comment:
I don't get the idea of recursing until we find `ConditionalExpressionSyntax`. Shouldn't we just register syntax node action for the needed syntax kind?

Reply:
This is better as it avoids all the callbacks and allows us to just bulk fix everything.

Reply:
@CyrusNajmabadi

> avoids all the callbacks

Did this show as perf issue with RegisterSyntaxNodeAction before? 

If there are, it's definitely better to improve the commonly used RegisterSyntaxNodeAction rather than traversing ourselves. 

Tagging @mavasani in case he can take a look at improving the API. 


> allows us to just bulk fix everything.

I don't quite understand this.

Reply:
> I don't quite understand this.

it allows the entire thing to be done in one swell foop.  No need for any callbacks at all.

Reply:
> Did this show as perf issue with RegisterSyntaxNodeAction before?

Yes.  I've had several times when measuring performance that i've seen callback-based walks over the tree be absolutely annihilated by a direct walk of the nodes.  It cannot be overstated just how fast that direct walk is.  Things like classification definitely revealed this.  My general mindset is to prefer that for these sorts of pure syntactic operations.